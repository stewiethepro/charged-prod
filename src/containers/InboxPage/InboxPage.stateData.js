import { bool, shape, string } from 'prop-types';
import {
  BOOKING_PROCESS_NAME,
  PURCHASE_PROCESS_NAME,
  resolveLatestProcessName,
  getProcess,
  isBookingProcess,
} from '../../transactions/transaction';

import { getStateDataForBookingProcess } from './InboxPage.stateDataBooking.js';
import { getStateDataForPurchaseProcess } from './InboxPage.stateDataPurchase.js';

export const stateDataShape = shape({
  processName: string.isRequired,
  processState: string.isRequired,
  actionNeeded: bool,
  isFinal: bool,
  isSaleNotification: bool,
});

// Translated name of the state of the given transaction
export const getStateData = params => {
  const { transaction } = params;
  const processName = resolveLatestProcessName(transaction?.attributes?.processName);
  const process = getProcess(processName);

  const processInfo = () => {
    const { getState, states } = process;
    const processState = getState(transaction);
    return {
      processName,
      processState,
      states,
    };
  };

  if (processName === PURCHASE_PROCESS_NAME) {
    return getStateDataForPurchaseProcess(params, processInfo());
  } else if (isBookingProcess(processName)) {
    return getStateDataForBookingProcess(params, processInfo());
  } else {
    return {};
  }

};
