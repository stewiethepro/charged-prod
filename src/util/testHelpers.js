import React from 'react';
import mapValues from 'lodash/mapValues';

import * as reactTestingLibrary from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { HelmetProvider } from 'react-helmet-async';

import configureStore from '../store';
import { IntlProvider } from './reactIntl';
import defaultConfig from '../config/configDefault';
import { ConfigurationProvider } from '../context/configurationContext';
import { RouteConfigurationProvider } from '../context/routeConfigurationContext';
import routeConfiguration from '../routing/routeConfiguration';

// In case you have translated the template and have new translations that
// are missing from the en translations file, the language for the tests can
// be changed here so that there are no missing translation keys in tests.
import messages from '../translations/en.json';

let undefined;
export const getDefaultConfiguration = () => {
  // TODO: add more relevant data for tests
  // TODO: make it possible to overwrite configuration for tests
  return {
    ...defaultConfig,
    currency: 'USD',
    marketplaceName: 'MarketplaceX',
    facebookAppId: undefined,
    marketplaceRootURL: 'http://localhost:3000',
    maps: {
      ...defaultConfig.maps,
      mapboxAccessToken: undefined,
      googleMapsAPIKey: undefined,
    },
    stripe: {
      ...defaultConfig.stripe,
      publishableKey: 'pk_test_',
    },
    listing: {
      listingFields: [
        {
          key: 'category',
          scope: 'public',
          includeForListingTypes: [
            'product-selling',
            'daily-booking',
            'nightly-booking',
            'hourly-booking',
          ],
          schemaType: 'enum',
          enumOptions: [
            { option: 'city-bikes', label: 'City bikes' },
            { option: 'e-bikes', label: 'E-bikes' },
            { option: 'mtb', label: 'MTB' },
            { option: 'kids-bikes', label: 'Kids bikes' },
          ],
          filterConfig: {
            indexForSearch: true,
            filterType: 'SelectSingleFilter',
            label: 'Category',
            group: 'primary',
          },
          showConfig: {
            label: 'Category',
            isDetail: true,
          },
          saveConfig: {
            label: 'Select category',
            placeholderMessage: 'Choose…',
            isRequired: true,
            requiredMessage: 'You need to select a category.',
          },
        },

        {
          key: 'tire-size',
          scope: 'public',
          includeForListingTypes: [
            'product-selling',
            'daily-booking',
            'nightly-booking',
            'hourly-booking',
          ],
          schemaType: 'enum',
          enumOptions: [
            { option: 29, label: '29' },
            { option: 28, label: '28' },
            { option: 27, label: '27' },
            { option: 26, label: '26' },
            { option: 24, label: '24' },
            { option: 20, label: '20' },
            { option: 18, label: '18' },
          ],
          filterConfig: {
            indexForSearch: true,
            label: 'Tire size',
            group: 'secondary',
          },
          showConfig: {
            label: 'Tire size',
            isDetail: true,
          },
          saveConfig: {
            label: 'Tire size',
            placeholderMessage: 'Choose…',
            isRequired: true,
            requiredMessage: 'You need to select a tire size.',
          },
        },
        {
          key: 'brand',
          scope: 'public',
          includeForListingTypes: [
            'product-selling',
            'daily-booking',
            'nightly-booking',
            'hourly-booking',
          ],
          schemaType: 'enum',
          enumOptions: [
            { option: 'cube', label: 'Cube' },
            { option: 'diamant', label: 'Diamant' },
            { option: 'ghost', label: 'GHOST' },
            { option: 'giant', label: 'Giant' },
            { option: 'kalkhoff', label: 'Kalkhoff' },
            { option: 'kona', label: 'Kona' },
            { option: 'otler', label: 'Otler' },
            { option: 'vermont', label: 'Vermont' },
          ],
          filterConfig: {
            indexForSearch: true,
            label: 'Brand',
            group: 'secondary',
          },
          showConfig: {
            label: 'Brand',
            isDetail: true,
          },
          saveConfig: {
            label: 'Brand',
            placeholderMessage: 'Choose…',
            isRequired: true,
            requiredMessage: 'You need to select a brand.',
          },
        },
      ],
      listingTypes: [
        {
          listingType: 'rent-bicycles',
          label: 'Rent bicycles',
          transactionType: {
            process: 'default-booking',
            alias: 'default-booking/release-1',
            unitType: 'day',
          },
        },
        {
          listingType: 'product-selling',
          label: 'Sell products',
          transactionType: {
            process: 'default-purchase',
            alias: 'default-purchase/release-1',
            unitType: 'item',
          },
        },
      ],
    },
    search: {
      ...defaultConfig.search,
      mainSearch: {
        searchType: 'keywords',
      },
      defaultFilters: [defaultConfig.search.dateRangeFilter, defaultConfig.search.priceFilter],
    },
  };
};

export const getRouteConfiguration = () => {
  const layoutConfig = {
    searchPage: { variantType: 'map' },
    listingPage: { variantType: 'carousel' },
    listingImage: {
      aspectWidth: 400,
      aspectHeight: 400,
      variantPrefix: 'listing-card',
    },
  };
  return routeConfiguration(layoutConfig);
};

// Locale should not affect the tests. We ensure this by providing
// messages with the key as the value of each message.
const testMessages = mapValues(messages, (val, key) => key);

// Provide all the context for components that connect to the Redux
// store, i18n, router, etc.
export const TestProvider = ({ children }) => {
  const store = configureStore();
  return (
    <ConfigurationProvider value={getDefaultConfiguration()}>
      <RouteConfigurationProvider value={getRouteConfiguration()}>
        <IntlProvider locale="en" messages={testMessages} textComponent="span">
          <Provider store={store}>
            <HelmetProvider>
              <MemoryRouter>{children}</MemoryRouter>
            </HelmetProvider>
          </Provider>
        </IntlProvider>
      </RouteConfigurationProvider>
    </ConfigurationProvider>
  );
};

// With react-testing-library we don't have option to shallow render
// so we would need a custom render function to add required wrappers
// in order to test the whole Page component.
// Example for creating custom render function
// for using React Intl with RTL:
// https://testing-library.com/docs/example-react-intl#creating-a-custom-render-function
// function render(ui, { preloadedState, store = configureStore(), ...renderOptions } = {}) {
//   function Wrapper({ children }) {
//     return (<TestProvider>{children}</TestProvider>);
//   }
//   return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
// }

export const renderWithProviders = (ui, renderOptions = {}) => {
  const Wrapper = ({ children }) => {
    return <TestProvider>{children}</TestProvider>;
  };
  return reactTestingLibrary.render(ui, { wrapper: Wrapper, ...renderOptions });
};
export const testingLibrary = { ...reactTestingLibrary, userEvent };
