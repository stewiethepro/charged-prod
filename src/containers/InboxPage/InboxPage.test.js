import React from 'react';
import '@testing-library/jest-dom';
import Decimal from 'decimal.js';

import { types as sdkTypes } from '../../util/sdkLoader';
import { LINE_ITEM_ITEM, LINE_ITEM_PROVIDER_COMMISSION } from '../../util/types';
import {
  createUser,
  createCurrentUser,
  createListing,
  fakeIntl,
  createTransaction,
} from '../../util/testData';
import { renderWithProviders as render, testingLibrary } from '../../util/testHelpers';

import {
  TX_TRANSITION_ACTOR_CUSTOMER,
  TX_TRANSITION_ACTOR_PROVIDER,
  getProcess,
} from '../../transactions/transaction';

import { getStateData } from './InboxPage.stateData';
import { InboxPageComponent, InboxItem } from './InboxPage';

const { Money } = sdkTypes;
const { screen, within } = testingLibrary;
const noop = () => null;

const transitions = getProcess('default-purchase')?.transitions;

describe('InboxPage', () => {
  const provider = createUser('provider-user-id');
  const customer = createUser('customer-user-id');
  const currentUserProvider = createCurrentUser('provider-user-id');
  const currentUserCustomer = createCurrentUser('customer-user-id');
  const listing = createListing('ItemX', {
    publicData: {
      listingType: 'sell-bikes',
      transactionProcessAlias: 'default-purchase',
      unitType: 'item',
    },
  });

  const lineItems = [
    {
      code: LINE_ITEM_ITEM,
      includeFor: ['customer', 'provider'],
      quantity: new Decimal(1),
      unitPrice: new Money(1000, 'USD'),
      lineTotal: new Money(1000, 'USD'),
      reversal: false,
    },
    {
      code: LINE_ITEM_PROVIDER_COMMISSION,
      includeFor: ['provider'],
      unitPrice: new Money(100 * -1, 'USD'),
      lineTotal: new Money(100 * -1, 'USD'),
      reversal: false,
    },
  ];

  const ordersProps = {
    location: { search: '' },
    history: {
      push: () => console.log('HistoryPush called'),
    },
    params: {
      tab: 'orders',
    },
    authInProgress: false,
    currentUser: currentUserProvider,
    currentUserHasListings: false,
    isAuthenticated: false,
    fetchInProgress: false,
    onLogout: noop,
    onManageDisableScrolling: noop,
    transactions: [
      createTransaction({
        id: 'order-1',
        lastTransition: transitions.CONFIRM_PAYMENT,
        customer,
        provider,
        listing,
        lastTransitionedAt: new Date(Date.UTC(2017, 0, 15)),
        lineItems,
      }),
      createTransaction({
        id: 'order-2',
        lastTransition: transitions.CONFIRM_PAYMENT,
        customer,
        provider,
        listing,
        lastTransitionedAt: new Date(Date.UTC(2016, 0, 15)),
        lineItems,
      }),
    ],
    intl: fakeIntl,
    scrollingDisabled: false,
    sendVerificationEmailInProgress: false,
    onResendVerificationEmail: noop,
  };

  const salesProps = {
    location: { search: '' },
    history: {
      push: () => console.log('HistoryPush called'),
    },
    params: {
      tab: 'sales',
    },
    authInProgress: false,
    currentUser: currentUserCustomer,
    currentUserHasListings: false,
    isAuthenticated: false,
    fetchInProgress: false,
    onLogout: noop,
    onManageDisableScrolling: noop,
    transactions: [
      createTransaction({
        id: 'sale-1',
        lastTransition: transitions.CONFIRM_PAYMENT,
        customer,
        provider,
        listing,
        lastTransitionedAt: new Date(Date.UTC(2017, 0, 15)),
        lineItems,
      }),
      createTransaction({
        id: 'sale-2',
        lastTransition: transitions.CONFIRM_PAYMENT,
        customer,
        provider,
        listing,
        lastTransitionedAt: new Date(Date.UTC(2016, 0, 15)),
        lineItems,
      }),
    ],
    intl: fakeIntl,
    scrollingDisabled: false,
    sendVerificationEmailInProgress: false,
    onResendVerificationEmail: noop,
  };

  test('InboxPageComponent has tabs and inbox items for orders', () => {
    render(<InboxPageComponent {...ordersProps} />);

    // Has links to orders and sales tabs
    const ordersTabTitle = screen.getByRole('link', { name: 'InboxPage.ordersTabTitle' });
    expect(ordersTabTitle).toBeInTheDocument();
    expect(ordersTabTitle.getAttribute('href')).toContain('/inbox/orders');

    const salesTabTitle = screen.getByRole('link', { name: 'InboxPage.salesTabTitle' });
    expect(salesTabTitle).toBeInTheDocument();
    expect(salesTabTitle.getAttribute('href')).toContain('/inbox/sales');

    // Has 2 items
    const items = screen.queryAllByRole('link', { name: /ItemX/i });
    expect(items).toHaveLength(2);

    const item1 = items[0];
    expect(item1.getAttribute('href')).toContain('/order/order-1');
    const status1 = within(item1).getByText('InboxPage.default-purchase.purchased.status');
    expect(status1).toBeInTheDocument();

    const item2 = items[1];
    expect(item2.getAttribute('href')).toContain('/order/order-2');
    const status2 = within(item2).getByText('InboxPage.default-purchase.purchased.status');
    expect(status2).toBeInTheDocument();
  });

  // This is quite small component what comes to rendered HTML
  // For now, we rely on snapshot-testing and checking quantity.
  test('InboxItem matches snapshot of order', () => {
    const stateDataOrder = getStateData({
      transaction: ordersProps.transactions[0],
      transactionRole: TX_TRANSITION_ACTOR_CUSTOMER,
    });

    const tree = render(
      <InboxItem
        tx={ordersProps.transactions[0]}
        transactionRole={TX_TRANSITION_ACTOR_CUSTOMER}
        intl={fakeIntl}
        stateData={stateDataOrder}
        isBooking={false}
        stockType="multipleItems"
      />
    );
    expect(tree.asFragment().firstChild).toMatchSnapshot();
    expect(screen.getByText('InboxPage.quantity')).toBeInTheDocument();
  });

  test('InboxPageComponent has tabs and inbox items for sales', () => {
    render(<InboxPageComponent {...salesProps} />);

    // Has links to orders and sales tabs
    const ordersTabTitle = screen.getByRole('link', { name: 'InboxPage.ordersTabTitle' });
    expect(ordersTabTitle).toBeInTheDocument();
    expect(ordersTabTitle.getAttribute('href')).toContain('/inbox/orders');

    const salesTabTitle = screen.getByRole('link', { name: 'InboxPage.salesTabTitle' });
    expect(salesTabTitle).toBeInTheDocument();
    expect(salesTabTitle.getAttribute('href')).toContain('/inbox/sales');

    // Has 2 items
    const items = screen.queryAllByRole('link', { name: /ItemX/i });
    expect(items).toHaveLength(2);

    const item1 = items[0];
    expect(item1.getAttribute('href')).toContain('/sale/sale-1');
    const status1 = within(item1).getByText('InboxPage.default-purchase.purchased.status');
    expect(status1).toBeInTheDocument();

    const item2 = items[1];
    expect(item2.getAttribute('href')).toContain('/sale/sale-2');
    const status2 = within(item2).getByText('InboxPage.default-purchase.purchased.status');
    expect(status2).toBeInTheDocument();
  });

  // This is quite small component what comes to rendered HTML
  // For now, we rely on snapshot-testing and checking quantity.
  test('InboxItem matches snapshot of sales', () => {
    const stateDataOrder = getStateData({
      transaction: salesProps.transactions[0],
      transactionRole: TX_TRANSITION_ACTOR_PROVIDER,
    });

    const tree = render(
      <InboxItem
        tx={salesProps.transactions[0]}
        transactionRole={TX_TRANSITION_ACTOR_PROVIDER}
        intl={fakeIntl}
        stateData={stateDataOrder}
        isBooking={false}
        stockType="oneItem"
      />
    );
    expect(tree.asFragment().firstChild).toMatchSnapshot();
    expect(screen.queryByText('InboxPage.quantity')).not.toBeInTheDocument();
  });
});
