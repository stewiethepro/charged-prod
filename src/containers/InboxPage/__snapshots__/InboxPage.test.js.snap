// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InboxPage InboxItem matches snapshot of order 1`] = `
<div
  class="item"
>
  <div
    class="itemAvatar"
  >
    <a
      class="root"
      href="/u/provider-user-id"
      title="provider-user-id display name"
    >
      <span
        class="initials"
      >
        TT
      </span>
    </a>
  </div>
  <a
    class="itemLink"
    href="/order/order-1"
  >
    <div
      class="rowNotificationDot"
    />
    <div
      class="itemUsername"
    >
      <span
        class=""
      >
        provider-user-id display name
      </span>
    </div>
    <div
      class="itemTitle"
    >
      ItemX title
    </div>
    <div
      class="itemDetails"
    >
      <span>
        InboxPage.quantity
      </span>
    </div>
    <div
      class="itemState"
    >
      <div
        class="stateName stateNoActionNeeded"
      >
        <span>
          InboxPage.default-purchase.purchased.status
        </span>
      </div>
    </div>
  </a>
</div>
`;

exports[`InboxPage InboxItem matches snapshot of sales 1`] = `
<div
  class="item"
>
  <div
    class="itemAvatar"
  >
    <a
      class="root"
      href="/u/customer-user-id"
      title="customer-user-id display name"
    >
      <span
        class="initials"
      >
        TT
      </span>
    </a>
  </div>
  <a
    class="itemLink"
    href="/sale/sale-1"
  >
    <div
      class="rowNotificationDot"
    >
      <div
        class="notificationDot"
      />
    </div>
    <div
      class="itemUsername"
    >
      <span
        class=""
      >
        customer-user-id display name
      </span>
    </div>
    <div
      class="itemTitle"
    >
      ItemX title
    </div>
    <div
      class="itemDetails"
    />
    <div
      class="itemState"
    >
      <div
        class="stateName stateActionNeeded"
      >
        <span>
          InboxPage.default-purchase.purchased.status
        </span>
      </div>
    </div>
  </a>
</div>
`;
